pkg/controller/utils/datadog/metrics_forwarder.go

```diff
@@ -0,0 +1,434 @@
+			mf.logger.Error(err, "cannot update DatadogAgentDeployment status")
+		}
+		return false, nil
+	}
+	mf.logger.Info("Initializing Datadog metrics forwarder")
+	if err := mf.initAPIClient(apiKey, appKey); err != nil {
+		mf.logger.Error(err, "cannot get Datadog metrics forwarder to send deployment metrics, will retry later...")
+		if err = mf.updateStatusIfNeeded(dad, err); err != nil {
+			mf.logger.Error(err, "cannot update DatadogAgentDeployment status")
+		}
+		return false, nil
+	}
+	if err = mf.updateStatusIfNeeded(dad, nil); err != nil {
+		mf.logger.Error(err, "cannot update DatadogAgentDeployment status")
+	}
+	return true, nil
+}
+
+// forwardMetrics sends metrics to Datadog
+// it tries to refresh credentials each time it's called
+// forwardMetrics updates status conditions of the Custom Resource
+// related to Datadog metrics forwarding by calling updateStatusIfNeeded
+func (mf *metricsForwarder) forwardMetrics() error {
+	dad, err := mf.getDatadogAgentDeployment()
+	if err != nil {
+		mf.logger.Error(err, "cannot get DatadogAgentDeployment to get deployment metrics")
+		return err
+	}
+	apiKey, appKey, err := mf.getCredentials(dad)
+	if err != nil {
+		mf.logger.Error(err, "cannot get Datadog credentials")
+		if err = mf.updateStatusIfNeeded(dad, err); err != nil {
+			mf.logger.Error(err, "cannot update DatadogAgentDeployment status")
+		}
+		return err
+	}
+	if err := mf.updateCredsIfNeeded(apiKey, appKey); err != nil {
+		mf.logger.Error(err, "cannot update Datadog credentials")
+		if err = mf.updateStatusIfNeeded(dad, err); err != nil {
+			mf.logger.Error(err, "cannot update DatadogAgentDeployment status")
+		}
+		return err
+	}
+	mf.logger.Info("Collecting metrics")
+	mf.updateTags(dad)
+	status := dad.Status.DeepCopy()
+	if err := mf.sendStatusMetrics(status); err != nil {
+		mf.logger.Error(err, "cannot send status metrics to Datadog")
+		if err = mf.updateStatusIfNeeded(dad, err); err != nil {
+			mf.logger.Error(err, "cannot update DatadogAgentDeployment status")
+		}
+		return err
+	}
+	return mf.updateStatusIfNeeded(dad, err)
+}
+
+// initAPIClient initializes and validates the Datadog API client
+func (mf *metricsForwarder) initAPIClient(apiKey, appKey string) error {
+	if mf.delegator == nil {
+		mf.delegator = mf
+	}
+	datadogClient, err := mf.validateCreds(apiKey, appKey)
+	if err != nil {
+		return err
+	}
+	mf.datadogClient = *datadogClient
+	mf.keysHash = hashKeys(apiKey, appKey)
+	return nil
+}
+
+// updateCredsIfNeeded used to update Datadog apiKey and appKey if they change
+func (mf *metricsForwarder) updateCredsIfNeeded(apiKey, appKey string) error {
+	if mf.keysHash != hashKeys(apiKey, appKey) {
+		return mf.initAPIClient(apiKey, appKey)
+	}
+	return nil
+}
+
+// validateCreds returns validates the creds by querying the Datadog API
+func (mf *metricsForwarder) validateCreds(apiKey, appKey string) (*api.Client, error) {
+	return mf.delegator.delegatedValidateCreds(apiKey, appKey)
```

Comment:
do we still need this `delegator` indirection? or can we call directly: 
```suggestion
	return mf.delegatedValidateCreds(apiKey, appKey)
```